google-oauth-client/src/main/java/com/google/api/client/auth/openidconnect/IdTokenVerifier.java

```diff
@@ -141,10 +194,60 @@ public final String getIssuer() {
    * @param idToken ID token
    * @return {@code true} if verified successfully or {@code false} if failed
    */
-  public boolean verify(IdToken idToken) {
-    return (issuers == null || idToken.verifyIssuer(issuers))
-        && (audience == null || idToken.verifyAudience(audience))
-        && idToken.verifyTime(clock.currentTimeMillis(), acceptableTimeSkewSeconds);
+  public boolean verify(IdToken idToken) throws VerificationException {
+    boolean simpleChecks =
+        (issuers == null || idToken.verifyIssuer(issuers))
+            && (audience == null || idToken.verifyAudience(audience))
+            && idToken.verifyTime(clock.currentTimeMillis(), acceptableTimeSkewSeconds);
+
+    if (!simpleChecks) {
+      return false;
+    }
+
+    if (Boolean.parseBoolean(environment.getVariable(SKIP_SIGNATURE_ENV_VAR))) {
+      return true;
+    }
+
+    // Short-circuit signature types
+    if (!SUPPORTED_ALGORITHMS.contains(idToken.getHeader().getAlgorithm())) {
+      throw new VerificationException(
+          "Unexpected signing algorithm: expected either RS256 or ES256");
+    }
+
+    PublicKey publicKeyToUse = null;
+    try {
+      String certificateLocation = getCertificateLocation(idToken.getHeader());
+      publicKeyToUse = publicKeyCache.get(certificateLocation).get(idToken.getHeader().getKeyId());
+    } catch (ExecutionException | UncheckedExecutionException e) {
+      throw new VerificationException("Error fetching PublicKey from certificate location", e);
+    }
+
+    if (publicKeyToUse == null) {
+      throw new VerificationException(
+          "Could not find PublicKey for provided keyId: " + idToken.getHeader().getKeyId());
+    }
+
+    try {
+      if (idToken.verifySignature(publicKeyToUse)) {
+        return true;
+      }
+      throw new VerificationException("Invalid signature");
+    } catch (GeneralSecurityException e) {
+      throw new VerificationException("Error validating token", e);
+    }
+  }
+
+  private String getCertificateLocation(Header header) throws VerificationException {
+    if (certificatesLocation != null) return certificatesLocation;
+
+    switch (header.getAlgorithm()) {
+      case "RS256":
+        return FEDERATED_SIGNON_CERT_URL;
+      case "ES256":
+        return IAP_CERT_URL;
+    }
+
+    throw new VerificationException("Unknown algorithm");
```

Comment:
Can you add the value of the algorithm in the exception? It would help troubleshooting.

Reply:
fixed, thx!